### PR TITLE
Add "include current period" for relative date filters and update filter layout

### PR DESCRIPTION
**Pull request by kdoh**
Implements #6443 and #4749 

TODOs: 

- [x] Conditionally show the control for "previous" and "next"
- [x] Options should use `Select` component with no offset
- [x] Tests for time-interval options map in backend code (@camsaul)
- [x] All of the states of this popover should be 380px wide, with the exception of the between state which is 550px (since it has two calendar pickers side by side).
- [x] The input box for specifying the previous/next days/weeks etc. should be right-aligned and much smaller, ~65px.
- [x] The height of the selects should match the height of the input boxes.
- [x] Select drop-downs should have minimum widths to accommodate the largest possible string so that the width of the popover doesn't jerk around.
- [x] As @salsakran already mentioned above, in the query builder, the rest of the filter options are cut off.
- [x] The include x so far string isn't correctly populating with the appropriate time granularity. It currently says include filters so far.
- [x] On dashboards, I expect the all options time filter widget to exactly match the date widget in the query builder, but it doesn't currently. It doesn't have all of the filtering options, and it doesn't have the include x so far checkbox.
- [x] On dashboards, the update filter button shouldn't extend the full width of the popover, but should be a standard purple button.
- [x] In the query builder, the filter popover doesn't remember my selection for including the current period or not. It's always checked, even if I uncheck it and update the filter via the button.
- [x] Trying to use the current period option with the Before filter results in problems:

### CODE REVIEW

**Comment by tlrobinson** [comment]
@kdoh I pushed a fix for "previous", "next", and "current" (i.e. `time-interval` filters).

Does this apply to the rest as well? Currently it's shown for all of them but I'm not sure that makes sense so I didn't implement it. Does it even make sense for the "current" filter?

**Comment by kdoh** [comment]
@tlrobinson Thanks. 

This should only apply to "previous" and "next", I hadn't gotten around to implementing the logic to only show the control in the case of those two specific `time-interval` filters. Would be awesome if you could help figure out the best way to do that.

The other thing at play here is the change to a horizontal layout. That's mostly done but one thing that feels like it should be cleaned up is the way we're handling the different options (Previous, Next, ... and also Day, Hour, etc.) Right now those are custom lists but it feels like they should be using our Select component. Only thing preventing that is we don't want to have them be our current popover style offset lists per the designs in #6443. Might be nice if the Select component had a "noOffset" or similar. 

I'll add a list of what I know is remaining in the PR body if you're able to pick those up.

**Comment by salsakran** [comment]
The filter operator box is being cut off 

screen shot 2017-12-19 at 5 53 38 pm

Also, it feels kind of weird to switch from vertical to horizontal layout, @kdoh  + @mazameli . It might feel smoother if we alter the dimensions a bit, not sure.

**Comment by mazameli** [comment]
I just took a look at this and noted a few to-dos, @tlrobinson:

1. All of the states of this popover should be 380px wide, with the exception of the `between` state which is 550px (since it has two calendar pickers side by side).
2. The input box for specifying the previous/next days/weeks etc. should be right-aligned and much smaller, ~65px.
3. The height of the selects should match the height of the input boxes.
4. Select drop-downs should have minimum widths to accommodate the largest possible string so that the width of the popover doesn't jerk around.
5. As @salsakran already mentioned above, in the query builder, the rest of the filter options are cut off.
6. The `include x so far` string isn't correctly populating with the appropriate time granularity. It currently says `include filters so far`.
7. On dashboards, I expect the `all options` time filter widget to exactly match the date widget in the query builder, but it doesn't currently. It doesn't have all of the filtering options, and it doesn't have the `include x so far` checkbox.
screen shot 2017-12-27 at 4 45 50 pm

8. On dashboards, the `update filter` button shouldn't extend the full width of the popover, but should be a standard purple button.

9. In the query builder, the filter popover doesn't remember my selection for including the current period or not. It's always checked, even if I uncheck it and update the filter via the button.

10. Trying to use the current period option with the Before filter results in problems:
screen shot 2017-12-27 at 4 56 48 pm

**Comment by tlrobinson** [comment]
I've fixed and cleaned up the logic for displaying and setting this option.

I haven't addressed any of the visual issues. Let me know if you'd like me to do that.

**Comment by mazameli** [comment]
@tlrobinson It would be awesome if you could address those visual issues. Thanks a ton.

**Comment by mazameli** [comment]
Happy new year, @tlrobinson. Let me know if and when you're ready for me to review changes; wasn't sure if you were working this week. Thanks!

**Comment by tlrobinson** [comment]
@mazameli I believe I've fixed all the visual issues in the qb. Please give it a try.

Regarding adding "include current period" to dashboards, the backend doesn't support that yet. We need to come up with a way for the frontend to serialize that bit of info in the URL parameter, and wire it up on the backend.

Is that in scope for this PR?

**Comment by mazameli** [comment]
This is looking awesome. It appears to be working correctly on dashboard filters when using the "all options" date widget, too. I can't find anything wrong with the UI/UX — bravo. 👏

**Comment by tlrobinson** [comment]
👍 

@metabase/core-developers Anyone want to review?

There's one item in the TODOs someone added for @camsaul, not sure if that's done. @camsaul?

**Comment by camsaul** [comment]
Yup I added tests when I added the dashboard Params support

File: src/metabase/query_processor/middleware/parameters/dates.clj [hunk 1]
```diff
@@ -109,12 +109,15 @@
                :end   (t/minus dt (t/days 1))})
     :filter (fn [_ field] ["=" field ["relative_datetime" -1 "day"]])}
 
-   {:parser (regex->parser #"past([0-9]+)(day|week|month|year)s", [:int-value :unit])
+   ;; adding a tilde (~) at the end of a past<n><unit> filter means we should include the current day/etc.
```
Comment (salsakran) [inline src/metabase/query_processor/middleware/parameters/dates.clj]: Can we have a test the makes sure this properly catches tildes?

File: frontend/src/metabase/components/Calendar.jsx [hunk 2]
```diff
@@ -109,7 +104,7 @@ export default class Calendar extends Component {
     renderWeeks(current) {
         var weeks = [],
             done = false,
-            date = moment(current).startOf("month").add("w" -1).day("Sunday"),
+            date = moment(current).startOf("month").day("Sunday"),
```
Comment (salsakran) [inline frontend/src/metabase/components/Calendar.jsx]: why did this remove the `add("w" -1)`?

**Review by salsakran — CHANGES_REQUESTED** [review]
Would be nice to test the expansion of the tilde.

File: src/metabase/query_processor/middleware/parameters/dates.clj (hunk 1 above)
Comment (camsaul) [reply src/metabase/query_processor/middleware/parameters/dates.clj]: We can ;) see https://github.com/metabase/metabase/pull/6530/files#diff-c775b10cbaf781a636a6e6090f2309a8R79

Comment (salsakran) [reply src/metabase/query_processor/middleware/parameters/dates.clj]: weird, I misread that test and never noticed the tilde

**Review by salsakran — APPROVED** [review]

**Comment by salsakran** [comment]
looks good, merging.

**Comment by MarkRx** [comment]
Running into an issue with this - #7339

**Pull request merged**